### ISSUE TITLE
When more than 1 pageful of uploads, -R still fails.

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/gothub#3]
cf. #1

**Comment by mfonville** [comment]
I have looked a bit through the documentation of the GitHub API.
And I think in GithubGet in api.go we should add processing of the Github response _headers_.
Because if there GitHub responds with a `Link:` field in its header, and they look like this:

```
Link: <https://api.github.com/repositories/35368260/releases/3190563/assets?access_token=XXXXXXXX&page=2>; rel="next", <https://api.github.com/repositories/35368260/releases/3190563/assets?access_token=XXXXXXXX&page=3>; rel="last"
```

So `rel="next"` indicates the next in-order pagination (on further pages the previous are indicated with `rel="prev"`) _if_ it exists (so not on the last page).
`rel="last"` gives the last page, also that one does _not_ exist on the last page itself. Also there is a `rel="first"` indication that is given. To visualize when i.e. fetching page3 of the example above:

```
Link: <https://api.github.com/repositories/35368260/releases/3190563/assets?access_token=XXXXXXXX&page=1>; rel="first", <https://api.github.com/repositories/35368260/releases/3190563/assets?access_token=XXXXXXXX&page=2>; rel="prev"
```

Proposal for the algorithm: if `rel="next"` exists, execute the function itself recursively with that URL. That should always result in fetching all available pages. The return values should be glued together when they are being returned.

**Comment by fasterthanlime** [comment]
@mfonville sounds good to me. Are you sure you don't want to give a try implementing it? I'll be out of town this week-end so it might be a week or more still until I have time to handle that.

Go tools are actually quite friendly :) they complain a lot but it prevents a lot of silly mistakes. I recommend using Atom, it's my editor of choice for Golang: http://marcio.io/2015/07/supercharging-atom-editor-for-go-development/

**Comment by mfonville** [comment]
@fasterthanlime I have literally zero experience with Go, and I looked at the code but wouldn't know where to start without knowing the basics.
I will look at the changes you make to the codebase and hope to get the hang of it, that's also how I learned Python within a few weeks :-)

There is no hurry until Android N is final, which is still a few months X-D

PS: Thanks for adopting this project and the current upload fix. It is great that @NicholasBuse and I don't have to attend to manually removing OpenGApps GitHub assets anymore.

**Comment by mfonville** [comment]
@fasterthanlime ok, Go is easier than I thought. PR might come still this afternoon :-D
But it needs more work to prevent pagination on situations where it is not necessary (i.e. when fetching just 1 tag, and has already been found on a page)

**Comment by NicholasBuse** [comment]
Yes, Thank you!!!

**Comment by mfonville** [comment]
@fasterthanlime I have a start: https://github.com/mfonville/gothub/commit/f4240b217c50f021ccbe968c032560eca859618e

But I don't get the recursion to work; either I use this `return` statement (line 129) before calling itself and it will return the last called page. If I do it without the `return` it only returns the first page.

~~I was reading https://www.goinggo.net/2013/09/recursion-and-tail-calls-in-go_26.html but I have no clue to account for that, because the `v interface {}` for passing the list the around is completely new to me and I have no clue how it works.~~

Found out that the interface is to mask those array structs, but still have no clue how to 'append' them together. Found this https://github.com/imdario/mergo but it is way above my head (I have no C background).

**Comment by mfonville** [comment]
@fasterthanlime would you be able to have a look at my current work and help me out with solving how to correctly call this function itself in a recursive way?

**Comment by fasterthanlime** [comment]
Looking at it now, I'm not sure `GithubGet` is the best place to implement pagination.

As you found out, `interface {}` is the "catch-all" type in golang, and then the JSON unmarshaller uses reflection to find out the actual type passed to `GithubGet`. There should probably be a variant that works on arrays instead - and we said fetching all pages unconditionally wasn't the best, right? What if it's in the first page?

**Comment by mfonville** [comment]
The problem is afaik if we'd do it in the parent functions of `GithubGet`, every function itself would have to become pagination-aware. Nicest way would be to do it function-agnostic, maybe an extra abstraction layer is necessary?

I indeed find out that it is the "catch-all", but especially the assignment style with `:=` to the function itself is hard to grasp me.

If the pagination should be not fetching 'unnecessary' pages it should maybe allow for specifying a needle to the `GithubGet`, but I am not sure how that'd work with such dynamic return types.
Currently I would argue that it is best to ask for the 100 results per page and all the pages available. That limits the amount of HTTP request (and counting toward the API query limit) and at least ensures a 100% reliable 'layer' for the logic that calls the gothub functions.

**Comment by fasterthanlime** [comment]
How about:
- Renaming `GithubGet` `GithubGetAdvanced`, add a new parameter, a struct that will be filled with the "prev" and "next" links if any
- Adding a `GithubGet` function that calls `GithubGetAdvanced` with a null third parameter
- Having the deletion code call `GithubGetAdvanced` and, as long as it doesn't find it, call it again (if there is a "next" link).

**Comment by mfonville** [comment]
That could work. But I'm thinking that there are also quite some other functions that probably would benefit by iterating over all pages. I.e. editing a release, or querying for all available releases would also all like to use pages if there are more than 100 of those.

The situation of actually having a 'needle' that can be checked and stopping the recursion when it is found, is only a subset of possible operations. Listing all available releases should e.g. never stop its querying preemptively.

How about this variation?
- A `GithubGetAll` or just keep `GithubGet` function that fetches all pages, used by default by most functions, that returns all the pages glued together. To do so it calls `GithubGetPage` until the 'next' result is empty.
- A `GithubGetPage` function that fetches a single page, returns both the results as the 'prev'+'next' when available.
- The deletion code can call the `GithubGetPage` directly instead of the All-variant, so that it can stop when the result has been found.

Even more nice, but I am not acquainted enough with how it would work in go, is to e.g. make the result an iterable that can be looped over until the desired result is found, and that the iterator would take care of calling the next page when applicable.

**Comment by fasterthanlime** [comment]
You proposed variation sounds okay to me!

**Comment by mfonville** [comment]
@fasterthanlime :-)
Now just need the coding of it, something I cannot unfortunately :'( (I contributed what I could, like the regex)

**Comment by mfonville** [comment]
Is there anything can do with my limited Go skills to help?

**Comment by fasterthanlime** [comment]
Should all be fixed in v0.7.0. Enjoy!

**Comment by mfonville** [comment]
@fasterthanlime thanks, you're the best! :-D
